docs(langsmith): update voice tracing guides for the langsmith.integrations SDK (#4636)
langchain-ai/docs: src/langsmith/trace-gemini-live.mdx

Reviewer: ```suggestion
sidebarTitle: Gemini Live
tag: Beta
```

description: Trace Gemini Live voice agents built with the Google Agent Development Kit (ADK) in LangSmith.
---

<Note>
This integration is in beta, so its API may change.
</Note>

Trace your [Gemini Live](https://ai.google.dev/gemini-api/docs/live-api) voice agents, built with the [Google Agent Development Kit (ADK)](https://google.github.io/adk-docs/streaming/), to LangSmith with the LangSmith ADK integration. For high-level conventions, see [Voice tracing fundamentals](/langsmith/trace-voice-fundamentals).

<Note>
The ADK Live integration requires `langsmith[google-adk-live]>=0.9.7` (separate from the `langsmith[google-adk]` batch integration).
</Note>

To trace text agents, tools, and multi-agent workflows built with ADK, see [Trace Google ADK applications](/langsmith/trace-with-google-adk).

> Gemini Live is a speech-to-speech model that ADK streams to your app as `run_live` events. The integration captures each conversation as a single LangSmith trace, with a span for every meaningful event (transcripts, tool calls, turn boundaries, and interruptions).

Reviewer: This sentence should move to the top of the page to be the intro sentence after the beta callout

## Install

<CodeGroup>

```bash pip
pip install "langsmith[google-adk-live]" google-genai
```

```bash uv

> uv add "langsmith[google-adk-live]" google-genai

Reviewer 2: Nice to have the version this was added in in a callout, e.g. https://docs.langchain.com/langsmith/trace-with-vercel-ai-sdk#ai-sdk-v5-and-v6

```

</CodeGroup>

## Set environment variables

```bash .env
LANGSMITH_API_KEY=<your-langsmith-api-key>
LANGSMITH_TRACING=true
LANGSMITH_PROJECT=<your-desired-langsmith-project>
GOOGLE_API_KEY=<your-google-api-key>
```

## Set up tracing

> Import `LangSmithGoogleADKLivePlugin` and register it on your `Runner`. It runs alongside your own `run_live` loop, so your loop keeps doing only audio playback and UI:

Reviewer: "keeps doing" what? I think we can use a more precise action here

```python
from google.adk.agents.run_config import RunConfig, StreamingMode
from google.adk.runners import Runner
from google.genai import types as genai_types
from langsmith.integrations.google_adk_live import LangSmithGoogleADKLivePlugin

plugin = LangSmithGoogleADKLivePlugin(project_name="gemini-live-voice")

runner = Runner(
    app_name="voice-app",
    agent=root_agent,
    session_service=session_service,
    plugins=[plugin],
)

run_config = RunConfig(
    response_modalities=["AUDIO"],
    streaming_mode=StreamingMode.BIDI,
    input_audio_transcription=genai_types.AudioTranscriptionConfig(),
    output_audio_transcription=genai_types.AudioTranscriptionConfig(),
)

async for event in runner.run_live(
    user_id=user_id,
    session_id=adk_session.id,
    live_request_queue=queue,
    run_config=run_config,
):
    ...  # play audio, handle barge-in, update the UI
```

<Note>

> Transcription is opt-in. Set both `input_audio_transcription` and `output_audio_transcription` on the `RunConfig`, so that the trace shows transcripts.

Reviewer: ```suggestion
Transcription is opt-in. To show transcripts, set both `input_audio_transcription` and `output_audio_transcription` on the `RunConfig`.
```

</Note>

<Note>

> On a graceful end (the live request queue closing), ADK fires its `after_run` callback and the plugin finalizes the trace for you. On a cancelled run, such as a console app that stops `run_live` on Ctrl-C, ADK may not fire that callback, so call `plugin.finalize(session_id=adk_session.id)` during teardown. It is idempotent and keyed by session id, so it is a no-op if ADK's callback already ran. Without it, the trace, including the audio attachment built at finalize, is lost.

Reviewer: ```suggestion
On a graceful end (the live request queue closing), the ADK sends its `after_run` callback and the plugin finalizes the trace for you.

On a cancelled run (for example, a console app that stops `run_live` on Ctrl-C), the ADK may not send that callback, call `plugin.finalize(session_id=adk_session.id)` during teardown; otherwise, you lose the trace and the audio attachment that is built at finalize. The call is idempotent, so it does nothing if ADK's callback already ran.
```

suggested rewording for concision and clarity

</Note>

## Group a conversation into a thread

> Each conversation is captured as its own trace with its own thread id. To supply your own id, for example to group the conversation with related interactions in a LangSmith [thread](/langsmith/threads), pass a `thread_id_provider` to the plugin:

Reviewer: ```suggestion
Each conversation is captured as its own trace with its own thread ID. To supply your own ID, for example to group the conversation with related interactions in a LangSmith [thread](/langsmith/threads), pass a `thread_id_provider` to the plugin:
```

```python
plugin = LangSmithGoogleADKLivePlugin(
    project_name="gemini-live-voice",
    thread_id_provider=lambda: thread_id,
)
```

> A single plugin instance is shared across every `run_live` call, and it resolves the thread id once per conversation, at its start. The default keeps concurrent conversations separate. If you pass a `thread_id_provider` on a server handling concurrent conversations, return the id for the current conversation rather than a fixed value, for example by reading a `ContextVar` set at the start of each run.

Reviewer 2: Fine I think for now but in the future we should have a general voice tracing guide and have this be part of that

Author: we have a general tracing guide, and we can def add this to it!

> A single plugin instance is shared across every `run_live` call, and it resolves the thread id once per conversation, at its start. The default keeps concurrent conversations separate. If you pass a `thread_id_provider` on a server handling concurrent conversations, return the id for the current conversation rather than a fixed value, for example by reading a `ContextVar` set at the start of each run.

Reviewer: ```suggestion
A single plugin instance is shared across every `run_live` call and it resolves the thread ID once, at the start of each conversation. The default keeps concurrent conversations separate. If you pass a `thread_id_provider` on a server handling concurrent conversations, it returns the ID for the current conversation rather than a fixed value; for example, by reading a `ContextVar` set at the start of each run.
```

suggested edits for clarity but not sure if this changes the meaning

## Record the conversation audio

> Feed the plugin your microphone and playback audio, and it attaches a single stereo recording (user left, agent right) to the trace:

Reviewer: ```suggestion
If you feed your microphone and playback audio to the plugin, it attaches a single stereo recording (user left, agent right) to the trace:
```

```python
plugin.record_user_audio(mic_chunk)      # user mic PCM16
plugin.record_agent_audio(played_chunk)  # agent PCM16 as played
```

> Record the user's microphone capture before resampling it for ADK, and tap the speaker for the agent's audio, so the recording reflects what was actually heard. Feed both channels at the same sample rate (the plugin's `sample_rate`, 24 kHz by default). For the underlying attachment API, see [Upload files with traces](/langsmith/upload-files-with-traces).

Reviewer 2: Maybe nice to add an example trace or screenshot somewhere

IIRC there was a 400 day expiry for shared traces but this could be lifted. Would ask platform folks

Author: Yep agreed. I'll do that in a follow up - don't want the screenshot being out of date too soon.

> Record the user's microphone capture before resampling it for ADK, and tap the speaker for the agent's audio, so the recording reflects what was actually heard. Feed both channels at the same sample rate (the plugin's `sample_rate`, 24 kHz by default). For the underlying attachment API, see [Upload files with traces](/langsmith/upload-files-with-traces).

Reviewer: ```suggestion
To accurately reflect what is heard, record the user's microphone capture before resampling it for ADK and tap the speaker for the agent's audio. Feed both channels at the same sample rate (the plugin's `sample_rate` is 24 kHz by default). For the underlying attachment API, see [Upload files with traces](/langsmith/upload-files-with-traces).
```
